src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseHeaderReader.cs

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Justin Van Patten. All rights reserved.
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+
+namespace System.Net.Http
+{
+    /// <summary>
+    /// Used to read header lines of an HTTP response, where each line is separated by "\r\n".
+    /// </summary>
+    internal struct WinHttpResponseHeaderReader
+    {
+        private const string Gzip = "gzip";
+        private const string Deflate = "deflate";
+
+        private readonly char[] _buffer;
+        private readonly int _length;
+        private int _position;
+
+        public WinHttpResponseHeaderReader(char[] buffer, int startIndex, int length)
+        {
+            CharArrayHelpers.DebugAssertArrayInputs(buffer, startIndex, length);
+
+            _buffer = buffer;
+            _position = startIndex;
+            _length = length;
+        }
+
+        /// <summary>
+        /// Reads a header line.
+        /// Empty header lines are skipped, as are malformed header lines that are missing a colon character.
+        /// </summary>
+        /// <returns>true if the next header was read successfully, or false if all characters have been read.</returns>
+        public bool ReadHeader(out string name, out string value)
+        {
+            int startIndex;
+            int length;
+            while (ReadLine(out startIndex, out length))
+            {
+                // Skip empty lines.
+                if (length == 0)
+                {
+                    continue;
+                }
+
+                int colonIndex = Array.IndexOf(_buffer, ':', startIndex, length);
+
+                // Skip malformed header lines that are missing the colon character.
+                if (colonIndex == -1)
+                {
+                    continue;
+                }
+
+                int nameLength = colonIndex - startIndex;
+
+                // If it's a known header name, use the known name instead of allocating a new string.
+                if (!HttpKnownHeaderNames.TryGetHeaderName(_buffer, startIndex, nameLength, out name))
```

Comment:
Can we use the same HttpKnownHeaderNames as an optimization in CurlHandler in a subsequent PR? (Same question for any other optimizations.)